/******************************************************************************
 *
 *  Copyright (c) 2014-2015, 2020-2021 The Linux Foundation. All rights reserved
 *  Copyright (C) 2003-2012 Broadcom Corporation
 *  Copyright (C) 2008, 2011-2012, 2016-2018 The Android Open Source Project
 *  Copyright (C) 2014 Google, Inc
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at:
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ******************************************************************************/

From cfb019d7b43a103dd1749f66c20de93554d5d1bf Mon Sep 17 00:00:00 2001
From: Sunny Kapdi <sunnyk@codeaurora.org>
Date: Sat, 13 Nov 2021 18:17:06 -0800
Subject: [PATCH] BT5.3: Add support for ADI in Periodic Advertising

Add support for ADI in Periodic Advertising. Ensure that the
Periodic Advertising Commands have checks for the support for ADI
in Controller.

Change-Id: I1c1dc675ddf0485215437f4668802b022803f7c4
---
 device/include/controller.h    | 1 +
 device/src/controller.cc       | 7 +++++++
 stack/btm/btm_ble_multi_adv.cc | 7 ++++++-
 stack/include/hcidefs.h        | 1 +
 4 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/device/include/controller.h b/device/include/controller.h
index 4410a6f1a7..670b397f06 100644
--- a/device/include/controller.h
+++ b/device/include/controller.h
@@ -66,6 +66,7 @@ typedef struct controller_t {
   bool (*supports_ble_periodic_advertising)(void);
   bool (*supports_ble_periodic_sync_transfer)(void);
   bool (*supports_ble_iso_broadcaster)(void);
+  bool (*supports_ble_periodic_advertising_adi)(void);
 
   // Get the cached acl data sizes for the controller.
   uint16_t (*get_acl_data_size_classic)(void);
diff --git a/device/src/controller.cc b/device/src/controller.cc
index a23f618fb9..595fcedcde 100644
--- a/device/src/controller.cc
+++ b/device/src/controller.cc
@@ -934,6 +934,12 @@ static bool supports_ble_iso_broadcaster(void) {
   return HCI_LE_ISO_BROADCASTER_SUPPORTED(features_ble.as_array);
 }
 
+static bool supports_ble_periodic_advertising_adi(void) {
+  CHECK(readable);
+  CHECK(ble_supported);
+  return HCI_LE_PERIODIC_ADVERTISING_ADI_SUPPORTED(features_ble.as_array);
+}
+
 static uint16_t get_acl_data_size_classic(void) {
   CHECK(readable);
   return acl_data_size_classic;
@@ -1196,6 +1202,7 @@ static const controller_t interface = {
     supports_ble_periodic_advertising,
     supports_ble_periodic_sync_transfer,
     supports_ble_iso_broadcaster,
+    supports_ble_periodic_advertising_adi,
 
     get_acl_data_size_classic,
     get_acl_data_size_ble,
diff --git a/stack/btm/btm_ble_multi_adv.cc b/stack/btm/btm_ble_multi_adv.cc
index 7bee9ea664..0a3bd97d0f 100644
--- a/stack/btm/btm_ble_multi_adv.cc
+++ b/stack/btm/btm_ble_multi_adv.cc
@@ -619,7 +619,7 @@ class BleAdvertisingManagerImpl
               return;
             }
 
-            c->self->SetPeriodicAdvertisingEnable(c->inst_id, true, Bind(
+            c->self->SetPeriodicAdvertisingEnable(c->inst_id, c->periodic_params.enable, Bind(
               [](c_type c, uint8_t status) {
                 if (!c->self) {
                   LOG(INFO) << "Stack was shut down";
@@ -958,6 +958,11 @@ class BleAdvertisingManagerImpl
         },
         p_inst, enable, std::move(cb));
 
+    if (enable != 0) {
+      if (!controller_get_interface()->supports_ble_periodic_advertising_adi()) {
+        enable = 1; // use value of 0x01 if ADI is not supported
+      }
+    }
     GetHciInterface()->SetPeriodicAdvertisingEnable(enable, inst_id,
                                                     std::move(enable_cb));
   }
diff --git a/stack/include/hcidefs.h b/stack/include/hcidefs.h
index 3e13e1bbb4..3b4a781d39 100644
--- a/stack/include/hcidefs.h
+++ b/stack/include/hcidefs.h
@@ -1528,6 +1528,7 @@ typedef struct {
 #define HCI_LE_CIS_SLAVE_SUPPORT(x) ((x)[3] & 0x20)
 #define HCI_LE_POW_CTR_REQ_SUPPORT(x) ((x)[4] & 0x02)
 #define HCI_LE_PATHLOSS_MONITORING_SUPPORT(x) ((x)[4] & 0x08)
+#define HCI_LE_PERIODIC_ADVERTISING_ADI_SUPPORTED(x) ((x)[4] & 0x10)
 
 /* Add_on features encoding - page 0 (the only page for now)*/
 #define HCI_WIPOWER_FASTBOOT_ENABLE(x) ((x)[0] & 0x01)
-- 
2.25.1

